### PR TITLE
Fix failure to sink last statements of `for` loops

### PR DESCRIPTION
**Pull request by zoffixznet**
**TL;DR: this fixes bugs but tanks performance of for loops by 8% and I'm scared to push it to master.**

-----

Fixes R#1531 https://github.com/rakudo/rakudo/issues/1531
Also fixes lack of explosion in `sub foo {fail}; foo for ^1`

In `for` loops, users would tend to expect last statement to get
sunk, but because our for loops are basically `.map` calls, they
don't. For methods calls, we already specially sink them if they're
the last statement, so `Failure.new for ^1` *does* blow up. But
we don't do anything for nqp::calls, which is why `sub foo {fail};
foo for ^1` doesn't blow up.

This patch fixes that problem and fixes stuff like $a Z+= 1 that
returns a Seq that actually needs to get sunk to perform the work.

There's nothing wrong with the patch as far as I can see, except
for one thing: it degrades performance by 8% when measured with
`sub foo {my $x}; { for ^14000_000 { foo }; say now - ENTER now }`
When I ran the stresstest, it too came out 8% slower than my earlier
runs. So... I'm kinda hesitant about pushing this to master.

### CODE REVIEW

**Comment by zoffixznet** [comment]
Actually, slower stresstest run [appears to have been a fluke](https://irclog.perlgeek.de/perl6-dev/2018-02-26#i_15860570) and more runs come out even 2s faster than on master, so doesn't look like it's as a massive perf loss as I originally thought.

Merging.
